library/core/src/intrinsics.rs

```diff
@@ -2470,6 +2470,25 @@ extern "rust-intrinsic" {
     /// constructing an empty slice) is returned.
     #[rustc_nounwind]
     pub fn option_payload_ptr<T>(arg: *const Option<T>) -> *const T;
+
+    /// Returns whether the argument is known at compile-time. This opens the
```

Comment:
What does this even mean? I don't see how this is a well-defined semantics.

You made this always return `true` in Miri which I highly doubt makes sense.

Reply:
Oh shoot - you're right, thought that was only for CTFE. That should be `false`

Reply:
The return value doesn't matter, both the `true` and the `false` code path should be the same, and just optimizations/perf changes. That said, returning `false` seems like the less weird option (tho we could also just pick a random path )

Reply:
No, always `false` is also clearly wrong, then the following code would be sound:
```rust
if is_constant(0) { unreachable_unchecked(); }
```

Reply:
I think the semantics is "non-deterministically return true or false" with the hope that the compiler picks "true" only if it knows the value of the argument.

So in Miri the result should be random. And in const-eval, we can only allow this once we are okay with allowing non-deterministic operations in `const`.

Reply:
> No, always `false` is also clearly wrong, then the following code would be sound:
> 
> ```rust
> if is_constant(0) { unreachable_unchecked(); }
> ```

I don't think so? A sufficiently dumb compiler could stick all constants into a local variable first and read it from that.

Reply:
> And in const-eval, we can only allow this once we are okay with allowing non-deterministic operations in const.

Disallowing it in a `const` context may be tricky due to how it's already used in `pow`. I don't fully see why we can't always return `true` for CTFE, though? Maybe you can explain further?

Reply:
We currently have the property that every `const fn` is deterministic when executed at runtime. Exposing `is_const` to `const fn` breaks that property.

Also see https://github.com/rust-lang/rfcs/pull/3352.

Reply:
I see, we should disallow it then. However, for rustc in particular, I think it makes sense to allow it *somehow* under the assumption that its usage is minimal and always functions the same. In the case of `pow`, the code always returns the same result no matter what, all I can think of that wouldn't be the same is the panic location on overflow

Reply:
We definitely cannot make this intrinsic stable. It should have a status similar to `const_eval_select`. I think we allow that to be used on a case-by-case basis, with the requirement that the externally observable behavior is the same no matter what `is_const` returns.

Reply:
Also AFAIK `const_eval_select` is unsafe, so the same should probably apply here. At least in a `const fn`.

Reply:
Since we don't inline in const eval, the result is technically always `false` for the use in `pow`. This intrinsic is not supposed to exist at runtime anymore, but instead the optimizer should replace it with true or false. That makes it a bit weird in MIR, as we interpret MIR. So we need to come up with our own rules for it

Reply:
It's not weird at all, it's a hint, like `black_box`: it as no semantics that can be relied upon by unsafe code, but it can steer the compiler in desired directions.

The compiler can also totally replace it by `rand()`, if it wants.

> the result is technically 

The result is technically non-deterministic. ;)

Reply:
We can freely choose if we want this to be true or false in const-eval. I have no idea which one makes more sense. But since the docs say `true` is a weak promise from the compiler that the value is known, `false` seems to me to be the safer default.

Reply:
`true` could also make sense since it is in const eval, and thus the value *is* known no matter what. I'm in favor of `false` though, basically using the "default path"

Reply:
No, `true` makes no sense. const-eval execution time corresponds to runtime, not compile-time.

Maybe `is_known_to_optimizer` would be a better name, and make it clear that returning `true` in CTFE would not make sense. But I think I prefer `is_compile_time_known`.

Reply:
*But* the machine has to know the value of `arg` at this point, anything else makes no sense as it's just interpreted the rest of the function and everything leading it. Thus it could then choose the branch to go down depending on whether the next checks succeed or not.

Is `true` a good idea, though? Not at all, this is for LLVM so doing so doesn't make sense. and since it won't even be allowed in `const fn`s there's not really much of a point in discussing it anyway.

Reply:
> _But_ the machine has to know the value of `arg` at this point, anything else makes no sense as it's just interpreted the rest of the function and everything leading it. Thus it could then choose the branch to go down depending on whether the next checks succeed or not.

Const eval is more like runtime execution than optimization. While it is happening at compile-time, the value being know at compile-time is just part of the weirdness of const eval. Const eval should behave like runtime behaves, so we need to pick the same value that runtime does (or start picking a "random" value, as either code path should be fine). We can make this new intrinsic behave as we want, but we need to figure out what we want.



> and since it won't even be allowed in `const fn`s there's not really much of a point in discussing it anyway.

don't you need that for the `pow` function, which is `const fn`?

Reply:
> don't you need that for the `pow` function, which is `const fn`?

We thankfully have `rustc_allow_const_fn_unstable` for that :)

Reply:
> But the machine has to know the value of arg at this point, anything else makes no sense as it's just interpreted the rest of the function and everything leading it.

When you run the program the regular way, by the time it reaches your `is_constant`, it also knows the value of the argument. So it's exactly the same as for const-eval. :) You can analyze programs statically and dynamically, and interpretation is definitely a dynamic technique, even if it occurs at compile-time. This intrinsic asks whether the value is statically known.